src/asInput/index.jsx

```diff
@@ -51,22 +55,36 @@ const asInput = (WrappedComponent, inputType = undefined, labelFirst = true) =>
       this.handleBlur = this.handleBlur.bind(this);
 
       const id = this.props.id ? this.props.id : newId('asInput');
+      const isValid = this.props.validator ? true : this.props.isValid;
+      const validationMessage = this.props.validator ? '' : this.props.validationMessage;
       this.state = {
         id,
         value: this.props.value,
-        isValid: true,
+        isValid,
+        validationMessage,
         describedBy: [],
         errorId: `error-${id}`,
         descriptionId: `description-${id}`,
       };
     }
 
     componentWillReceiveProps(nextProps) {
+      const updatedState = {};
       if (nextProps.value !== this.props.value) {
-        this.setState({
-          value: nextProps.value,
-        });
+        updatedState.value = nextProps.value;
       }
+      if (nextProps.isValid !== this.props.isValid && !nextProps.validator) {
+        updatedState.isValid = nextProps.isValid;
+      }
+      if (nextProps.validationMessage !== this.props.validationMessage && !nextProps.validator) {
+        updatedState.validationMessage = nextProps.validationMessage;
+      }
+      // If validator goes away, revert to props
+      if (nextProps.validator !== this.props.validator && !nextProps.validator) {
+        updatedState.isValid = nextProps.isValid;
+        updatedState.validationMessage = nextProps.validationMessage;
+      }
+      this.setState(updatedState);
```

Comment:
Nit: probably don't need to trigger a re-render if `updateState` is empty (I'm assuming `this.setState` is not smart like that, but it very well could be).

Reply:
Alright, I'm looking at code for this. React had better not be that dumb.

Reply:
Ugh it is that dumb. OK, will fix.

Reply:
Fixed.